File: Stack/P224BasicCalculator/BasicCalculator.py
import unittest

class Solution:
    def calculate(self, s: str) -> int:
        stack = []
        s = list(s)
        res, sign, num = 0, 1, 0
        for idx, val in enumerate(s):
            if val.isdigit():
                num = 10*num + int(val)
            elif val == '+':
                res += num * sign
                sign = 1
                num = 0
            elif val == '-':
                res += num * sign
                sign = -1
                num = 0
            elif val == '(':
                # store the previous result of expression into stack
                stack.append(res)
                stack.append(sign)
                sign = 1
                # a new expression, at this time, res resets to zero
                res = 0
                num = 0
            elif val == ')':
                res += num * sign   # 计算括号内的expression, 
                res *= stack.pop()  # sign
                res += stack.pop()  # num
                sign = 1
                num = 0
        res += sign * num
        return res

class UnitTest(unittest.TestCase):
    def test_1(self):
        s = "(1+(4+5+2)-3)+(6+8)"
        self.assertEqual(23, Solution().calculate(s))

if __name__ == "__main__":
    unittest.main()

File: Stack/P224BasicCalculator/BasicCalculator.java
package Stack.P224BasicCalculator;

import java.util.Stack;

/**
 * @CLassName BasicCalculator
 * @Description TODO
 * @Author cst
 * @Date 1/12/21  11:01 PM
 */
public class BasicCalculator {
    private static BasicCalculator ONLY;

    public static BasicCalculator getONLY() {
        if (ONLY == null) {
            ONLY = new BasicCalculator();
        }
        return ONLY;
    }

    public int calculate(String s) {
        char[] ch = s.toCharArray();
        Stack<Integer> stack = new Stack<>();
        int sign = 1, num = 0, res = 0;
        for (int i = 0; i < ch.length; i++) {
            if (Character.isDigit(ch[i])) {
                num = num * 10 + ch[i] - '0';
            } else if (ch[i] == '+') {
                res += num * sign;
                sign = 1;
                num = 0;
            } else if (ch[i] == '-') {
                res += num * sign;
                sign = -1;
                num = 0;
            } else if (ch[i] == '(') {
                stack.push(res);
                stack.push(sign);
                sign = 1;
                num = 0;
                res = 0;
            } else if (ch[i] == ')') {
                res += num * sign;
                res *= stack.pop();
                res += stack.pop();
                sign = 1;
                num = 0;
            }
        }
        res += num * sign;
        return res;
    }
}
